### PR TITLE
[FEAT] Direct estimation probability two random records match

### LINKED ISSUE #462: Estimating proportion of matches inaccurate, especially for small datasets
[Reproducible example](https://github.com/moj-analytical-services/splink_demos/blob/f48aeaccf77389ceba03fa844526c1d7edfdad68/quickstart_demo_deduplication.ipynb)

Particularly on small datasets, the proportion of matches is often greatly overestimated.

You can see the calculations by turning on logging to level 15:
```

logging.basicConfig(
    format="%(message)s",
)
splink_logger = logging.getLogger("splink")
splink_logger.setLevel(15)
```

And you'll see output like:

<details>

```
----- Starting EM training session -----

Original proportion of matches: 0.004
Applying comparison dob using bayes factor 171.000

Proportion of matches adjusted for blocking on l.dob = r.dob: 0.407
Training the m probabilities of the model by blocking on:
l.dob = r.dob

Parameter estimates will be made for the following comparison(s):
    - first_name
    - surname
    - city
    - email

Parameter estimates cannot be made for the following comparison(s) since they are used in the blocking rules: 
    - dob

Iteration 1: Largest change in params was 0.291 in proportion_of_matches
Iteration 2: Largest change in params was 0.173 in proportion_of_matches
Iteration 3: Largest change in params was 0.0587 in proportion_of_matches
Iteration 4: Largest change in params was 0.0212 in proportion_of_matches
Iteration 5: Largest change in params was 0.00951 in proportion_of_matches
Iteration 6: Largest change in params was 0.00493 in proportion_of_matches
Iteration 7: Largest change in params was 0.0028 in proportion_of_matches
Iteration 8: Largest change in params was 0.00167 in proportion_of_matches
Iteration 9: Largest change in params was 0.00104 in proportion_of_matches
Iteration 10: Largest change in params was 0.000659 in proportion_of_matches

EM converged after 10 iterations

Proportion of matches from trained model blocking on l.dob = r.dob: 0.971
Reversing comparison level dob using bayes factor 171.000
This estimate of proportion of matches now:  0.165

Median of prop of matches estimates: 0.165

Your model is not yet fully trained. Missing estimates for:
    - dob (no m values are trained).

----- Starting EM training session -----

Original proportion of matches: 0.165
Applying comparison first_name using bayes factor 74.807
Applying comparison surname using bayes factor 63.265

Proportion of matches adjusted for blocking on l.first_name = r.first_name and l.surname = r.surname: 0.999
Training the m probabilities of the model by blocking on:
l.first_name = r.first_name and l.surname = r.surname

Parameter estimates will be made for the following comparison(s):
    - dob
    - city
    - email

Parameter estimates cannot be made for the following comparison(s) since they are used in the blocking rules: 
    - first_name
    - surname

Iteration 1: Largest change in params was 0.23 in the m_probability of dob, level `All other comparisons`
Iteration 2: Largest change in params was -0.00333 in proportion_of_matches
Iteration 3: Largest change in params was -0.00304 in proportion_of_matches
Iteration 4: Largest change in params was -0.00258 in proportion_of_matches
Iteration 5: Largest change in params was -0.00206 in proportion_of_matches
Iteration 6: Largest change in params was -0.00157 in proportion_of_matches
Iteration 7: Largest change in params was -0.00115 in proportion_of_matches
Iteration 8: Largest change in params was -0.000818 in proportion_of_matches

EM converged after 8 iterations

Proportion of matches from trained model blocking on l.dob = r.dob: 0.971
Reversing comparison level dob using bayes factor 166.048
This estimate of proportion of matches now:  0.169

Proportion of matches from trained model blocking on l.first_name = r.first_name and l.surname = r.surname: 0.972
Reversing comparison level first_name using bayes factor 74.807
This estimate of proportion of matches now:  0.314
Reversing comparison level surname using bayes factor 63.265
This estimate of proportion of matches now:  0.007

Median of prop of matches estimates: 0.088

Your model is fully trained. All comparisons have at least one estimate for their m and u values
```
</details>

In this model, the true proportion of matches is 0.00695

But here’s an example of the problem:
The dataset is just 1,000 records
There are 394 distinct values for date of birth, which suggests the bayes factor on DOB is (at most) 394
But in fact in this dataset, a match of DOB always implies a true match.  Because in reality, the bayes factor no dob is much higher (cardinality much higher than 394, it’s just this is not seen in the data because it’s a small dataset) (edited) 

So:
- Splink undercounts the probability of match amongst matches on DOB - it thinks its 97%.  Because the match weights are (generally) too low, because the true cardinality of variables is underestimated.  (Note the 97% here is being driven mostly by the non dob match weights)
- When we ‘reverse out’ dob to get the global proportion of matches, this results in a proportion of matches estimate that’s too high.
- I don’t think there’s any reason these two effects should balance one another out

Another interesting observation though:
On this model, the data is small enough to compute the proportion of matches from the cartesian product:
```
# Blocking_rules_for_training is blank i.e. cartesian join
df_predict = linker.predict()

sql = f"""
select sum(match_probability)/(select count(*) from {df_predict.physical_name}) as proportion_of_matches
from {df_predict.physical_name}

"""
linker.con.execute(sql).fetch_df()
>0.009199
```

which is pretty close to being right

### PR DESCRIPTION
**Pull request by RobinL**
Closes #462.  See #462 for a description of why this functionality is an improvement

### CODE REVIEW

**Comment by RobinL** [comment]
image

**Comment by RobinL** [comment]
[Robin Linacre](https://app.slack.com/team/U01FJAR5891)
:no_entry:  [3 minutes ago](https://mojdt.slack.com/archives/C02TCQLLJAX/p1661422970995849?thread_ts=1661343607.765229&cid=C02TCQLLJAX)
Comparing this to the previous estimation methodology:
True value:  1/4,207
Blocking on first name and surname: estimate gλ = 1/152,166
Blocking on dob: estimate gλ = 1/4,781
Overall estimate from EM approach (median):  gλ = 39,305


[Robin Linacre](https://app.slack.com/team/U01FJAR5891)
:no_entry:  [2 minutes ago](https://mojdt.slack.com/archives/C02TCQLLJAX/p1661423003958089?thread_ts=1661343607.765229&cid=C02TCQLLJAX)
New methodology in this case (which is a particularly tricky one):   gλ: 1/7,661
So new methodology is a big improvement (edited)

**Comment by RobinL** [comment]
Equivalent for the 1,000 row demo in splink_demos:
True value gλ = 1/246
New technique = 1/341
Old technique:
Estimate 1:  1/778
Estimate 2:  1/16
Median = 1/103
So new technique wins again.
This is another very poor quality dataset, so in a more realistic example, new technique would do even better

File: docs/linkerest.md [hunk 1]
```diff
@@ -5,10 +5,19 @@ tags:
```
Comment (RobinL) [inline docs/linkerest.md:5]: Note this file was previously saved incorrectly.  This fix is nothing to do with this PR

File: splink/linker.py [hunk 2]
```diff
@@ -1919,3 +1924,50 @@ def load_settings_from_json(self, in_path: str):
...
+            deterministic_matching_rules (list): A list of deterministic matching
+                rules that should be designed to admit very few (none if possible)
+                false positives
+            recall (float): A guess at the recall the deterministic matching rules
+                will attain.  i.e. what proportion of true matches will be recovered
+                by these deterministic rules
+        """
+
+        # If user, by error, provides a single rule as a string
+        if isinstance(deterministic_matching_rules, str):
```
Comment (ThomasHepworth) [inline splink/linker.py:1947]: You could also just use [`ensure_is_list`](https://github.com/moj-analytical-services/splink/blob/master/splink/misc.py#L44) here

**Pull request merged**